incubator/hnc/internal/reconcilers/object.go

```diff
@@ -274,14 +274,16 @@ func (r *ObjectReconciler) skipNamespace(ctx context.Context, log logr.Logger, i
 // which indicates that we need to call the regular syncObject method. Note that this method may
 // modify `inst`.
 func (r *ObjectReconciler) syncMissingObject(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) syncAction {
-	// If the object exists, skip
+	// If the object exists, skip.
 	if inst.GetCreationTimestamp() != (v1.Time{}) {
 		return actionUnknown
 	}
 
+	ns := r.Forest.Get(inst.GetNamespace())
 	// If it's a source, it must have been deleted. Update the forest and enqueue all its
 	// descendants, but there's nothing else to do.
-	if r.forestHasSource(inst) {
+	if ns.HasSourceObject(r.GVK, inst.GetName()) {
+		ns.DeleteSourceObject(r.GVK, inst.GetName())
 		r.syncUnpropagatedSource(ctx, log, inst)
```

Comment:
It has reconcile HNCConfig and `enqueueDescendants()` to delete the obsolete copies.

Reply:
No, I just mean inline its features into this function since it does so little.

Reply:
`syncUnpropagatedSource` func is used twice and it contains two parts. I feel better to keep this logic stand alone, WDYT?
```
// syncUnpropagatedSource enqueues any propagated copies of the source.
//
// The method can be called when the source was deleted by users, or if it will no longer be
// propagated, e.g. because the user's changed the sync mode in the HNC Config.
func (r *ObjectReconciler) syncUnpropagatedSource(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) {
	// Signal the config reconciler for reconciliation because it is possible that the source object is
	// deleted on the apiserver.
	hnccrSingleton.requestReconcile("possible deleted source object")

	r.enqueueDescendants(ctx, log, inst, "possibly deleted source object")
}
```

Reply:
Ah sgtm